### ISSUE TITLE
The backward of 'bilinear Interpolation' is undeterministic

### ISSUE THREAD

**Issue by JingChaoLiu** [open-mmlab/mmsegmentation#255]
**Checklist**

1. I have searched related issues but cannot get the expected help.
2. The bug has not been fixed in the latest version.

**Describe the bug**
The backward of 'bilinear Interpolation' is undeterministic. To expand, once a model contains `torch.nn.functional.interpolate` with `mode={linear, bilinear, bicubic}`, the following settings cannot guarantee a reproducible results. 
```
random.seed(seed)
os.environ['PYTHONHASHSEED'] = str(seed)
np.random.seed(seed)
torch.manual_seed(seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
```
As mentioned in the issue [RFC: Add torch.deterministic flag to force deterministic algorithms](https://github.com/pytorch/pytorch/issues/15359) on **pytorch** official repository even in pytorch 1.7, this problem still haven't been solved(2020.11.13). The last reply of official developer is painted with yellow and green.

![image](https://user-images.githubusercontent.com/29098261/99047113-abc16100-25ce-11eb-8233-ab2fd851117f.png)

**Reproduction**
Using the above code snippet to fix the random seed and keeping all settings the same, we train deeplabv3+ (containing bilinear interpolation) twice on the cityscapes dataset, but the mIoU difference between the two models on the validation set is 0.7. The randomness of the experimental results even made it impossible for me to do ablation study safely in my paper.

After replacing the bilinear interpolation with the nearest neighbor interpolation, all the problems are solved, and even I can ensure that the printed losses are consistent within two repeated experiments from `iter=0` to `iter=40000`.

**Environment**
`{cuda=10.2, pytorch=1.6.0, mmseg=0.6.0}`

**Bug fix**
Replace `bilinear_interpolation(scale_factor)` with `nearest_interpolation(scale_factor) + conv(kernel_size=scale_factor+1)` can make the model to be reproducible and obtain a similar performance.
Implementation:
```
class Interpolate(nn.Module):

    def __init__(self, channel: int, scale_factor: int):
        super().__init__()
        # assert 'mode' not in kwargs and 'align_corners' not in kwargs and 'size' not in kwargs
        assert isinstance(scale_factor, int) and scale_factor > 1 and scale_factor % 2 == 0
        self.scale_factor = scale_factor
        kernel_size = scale_factor + 1  # keep kernel size being odd
        self.weight = nn.Parameter(
            torch.empty((1, 1, kernel_size, kernel_size), dtype=torch.float32).expand(channel, -1, -1, -1)
        )
        self.conv = functools.partial(
            F.conv2d, weight=self.weight, bias=None, padding=scale_factor // 2, groups=channel
        )
        with torch.no_grad():
            self.weight.fill_(1 / (kernel_size * kernel_size))

    def forward(self, t):
        if t is None:
            return t
        return self.conv(F.interpolate(t, scale_factor=self.scale_factor, mode='nearest'))

    @staticmethod
    def naive(t: torch.Tensor, size: Tuple[int, int], **kwargs):
        if t is None or t.shape[2:] == size:
            return t
        else:
            assert 'mode' not in kwargs and 'align_corners' not in kwargs
            return F.interpolate(t, size, mode='nearest', **kwargs)
```
especially, `Interpolate.naive` is designed for `Tensor(B, C, H=1, W=1) -> Tensor(B, C, *size)`.

**Comment by xvjiarui** [comment]
Hi @JingChaoLiu 
You are right. But replacing `bilinear_interpolation(scale_factor)` with `nearest_interpolation(scale_factor) + conv(kernel_size=scale_factor+1)` could not guarantee the original behavior. 

Nevertheless, your proposal is already supported [here](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/utils/up_conv_block.py).

**Comment by JingChaoLiu** [comment]
I think the code repository should maintain the reproducibility of the code. The randomness caused by bilinear interpolation (in my case, mIoU fluctuates by 0.7) prevents me from safely performing ablation experiments. I am afraid that this is not the original intention of why the benchmark was built.

**Comment by xvjiarui** [comment]
We are not sure if there  will degenerate performance. And this will be the BC breaking change. 
We may consider this in the next major release.

**Comment by JingChaoLiu** [comment]
Many Thanks. Hope more people can realize "bilinear interpolation is undeterministic" before being pointed out by reviewers at the rebuttal stage like me.
